src/analytics/private/segments/user-segments-workflow.ts

```diff
@@ -0,0 +1,217 @@
+ *
+ *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
+ *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
+ *  and limitations under the License.
+ */
+
+import { join } from 'path';
+import { Aws, Duration } from 'aws-cdk-lib';
+import { ISecurityGroup, IVpc, SubnetSelection } from 'aws-cdk-lib/aws-ec2';
+import { IRole, PolicyStatement } from 'aws-cdk-lib/aws-iam';
+import { IFunction } from 'aws-cdk-lib/aws-lambda';
+import { RetentionDays } from 'aws-cdk-lib/aws-logs';
+import { Bucket } from 'aws-cdk-lib/aws-s3';
+import {
+  Choice,
+  Condition,
+  Fail,
+  IStateMachine,
+  LogLevel,
+  StateMachine,
+  Succeed,
+  TaskInput,
+  Wait,
+  WaitTime,
+} from 'aws-cdk-lib/aws-stepfunctions';
+import { LambdaInvoke } from 'aws-cdk-lib/aws-stepfunctions-tasks';
+import { Construct } from 'constructs';
+import { SegmentJobStatus } from './segments-model';
+import { createLambdaRole } from '../../../common/lambda';
+import { createLogGroup } from '../../../common/logs';
+import { REDSHIFT_MODE } from '../../../common/model';
+import { SolutionNodejsFunction } from '../../../private/function';
+import { ExistingRedshiftServerlessProps, ProvisionedRedshiftProps } from '../model';
+
+export interface UserSegmentsWorkflowProps {
+  readonly projectId: string;
+  readonly networkConfig: {
+    readonly vpc: IVpc;
+    readonly vpcSubnets: SubnetSelection;
+  };
+  readonly securityGroupForLambda: ISecurityGroup;
+  readonly clickstreamMetadataDdbArn: string;
+  readonly dataAPIRole: IRole;
+  readonly serverlessRedshift?: ExistingRedshiftServerlessProps;
+  readonly provisionedRedshift?: ProvisionedRedshiftProps;
+  readonly databaseName: string;
+  readonly pipelineS3Bucket: string;
+  readonly segmentsS3Prefix: string;
+}
+
+const lambdaRootPath = __dirname + '/../../lambdas/user-segments-workflow';
+
+export class UserSegmentsWorkflow extends Construct {
+  private readonly props: UserSegmentsWorkflowProps;
+  public readonly userSegmentsWorkflow: IStateMachine;
+
+  constructor(scope: Construct, id: string, props: UserSegmentsWorkflowProps) {
+    super(scope, id);
+
+    this.props = props;
+    this.userSegmentsWorkflow = this.createWorkflow(props);
+  }
+
+  private createWorkflow(props: UserSegmentsWorkflowProps): IStateMachine {
+    // Define task for segment job initialization
+    const segmentJobInitTask = new LambdaInvoke(this, 'WorkflowTask-SegmentJobInit', {
+      lambdaFunction: this.constructNodejsFunction('segment-job-init', [
+        new PolicyStatement({
+          actions: [
+            'dynamodb:GetItem',
+            'dynamodb:PutItem',
+          ],
+          resources: [props.clickstreamMetadataDdbArn],
+        }),
+        new PolicyStatement({
+          actions: ['events:DisableRule'],
+          resources: [`arn:${Aws.PARTITION}:events:*:${Aws.ACCOUNT_ID}:rule/*`],
```

Comment:
Could we limit the rule name with prefix?

Reply:
Add `Clickstream-` prefix constraint